File: app/src/main/java/com/yenvth/soilDetectionApp/map/MapActivity.java
package com.yenvth.soilDetectionApp.map;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.UiSettings;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.maps.android.data.geojson.GeoJsonFeature;
import com.google.maps.android.data.geojson.GeoJsonLayer;
import com.google.maps.android.data.geojson.GeoJsonPolygonStyle;

import com.makeramen.roundedimageview.RoundedImageView;
import com.yenvth.soilDetectionApp.R;
import com.yenvth.soilDetectionApp.models.ProvinceModel;
import com.yenvth.soilDetectionApp.models.SoilModel;
import com.yenvth.soilDetectionApp.soilDetail.SoilDetailActivity;
import com.yenvth.soilDetectionApp.sqliteHelper.DBHelper;
import com.yenvth.soilDetectionApp.utils.CommonUtils;
import com.yenvth.soilDetectionApp.utils.Constant;
import com.yenvth.soilDetectionApp.utils.FileUtils;

import net.cachapa.expandablelayout.ExpandableLayout;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MapActivity extends AppCompatActivity implements
        MapView,
        OnMapReadyCallback,
        View.OnClickListener,
        GoogleMap.OnMapClickListener,
        GoogleMap.OnMarkerClickListener,
        MapAdapter.OnSoilMapItemClickListener{

    @BindView(R.id.btnBack)
    protected ImageView btnBack;
    @BindView(R.id.recycler_view_soil_map)
    protected RecyclerView recycler_view_soil_map;
    @BindView(R.id.expandLayout)
    protected ExpandableLayout expandLayout;
    @BindView(R.id.expandDetail)
    protected ExpandableLayout expandDetail;
    @BindView(R.id.tvDetail)
    protected TextView tvDetail;
    @BindView(R.id.spProvince)
    protected Spinner spProvince;
    @BindView(R.id.imgSoil)
    protected RoundedImageView imgSoil;
    @BindView(R.id.tvName)
    protected TextView tvName;
    @BindView(R.id.tvProvince)
    protected TextView tvProvince;
    @BindView(R.id.tvDes)
    protected TextView tvDes;
    @BindView(R.id.progressBar)
    protected ProgressBar progressBar;

    private FusedLocationProviderClient fusedLocationProviderClient;
    private GoogleMap googleMap;
    private UiSettings mUiSettings;
    private MapPresenterImpl<MapView> presenter;

    private MapAdapter mAdapter;
    private ArrayList<SoilModel> soilModels;
    private SoilModel soilSelected;

    private DBHelper dbHelper;
    private ArrayList<ProvinceModel> provinceModels;
    private String provinceSelected;
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_map);
        getSupportActionBar().hide();
        ButterKnife.bind(this);
        presenter = new MapPresenterImpl<>(this, this);
        init();
        action();
    }

    private void init() {
        dbHelper = new DBHelper(this);
        provinceModels = dbHelper.getProvinces();

        SupportMapFragment mapFragment =
                (SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);
        presenter.getListSoil("");
    }

    private void action() {
        btnBack.setOnClickListener(this);
        tvDetail.setOnClickListener(this);
        //Todo: Setup spinner
        ArrayAdapter provinceAdapter = new ArrayAdapter<>(this, R.layout.layout_custom_spinner_not_icon, provinceModels);
        provinceAdapter.setDropDownViewResource(R.layout.item_spinner);
        spProvince.setAdapter(provinceAdapter);

        spProvince.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                provinceSelected = provinceModels.get(i).getProvince_name();
                CommonUtils.showError(MapActivity.this, provinceSelected);
                presenter.getListSoil("");
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
    }



    @Override
    public void onMapReady(GoogleMap googleMap) {
        this.googleMap = googleMap;
        moveCameraMap(Constant.MapEnum.HANOT_LAT, Constant.MapEnum.HANOT_LON);
        googleMap.setOnMapClickListener(this);

        if (mUiSettings != null) {
            mUiSettings = googleMap.getUiSettings();
            mUiSettings.setZoomControlsEnabled(false);
            mUiSettings.setMyLocationButtonEnabled(true);
            mUiSettings.setZoomGesturesEnabled(false);
            mUiSettings.setTiltGesturesEnabled(false);
            mUiSettings.setRotateGesturesEnabled(false);
            mUiSettings.setMapToolbarEnabled(false);
            mUiSettings.setScrollGesturesEnabled(false);
        }
        try {
            addGeoJsonLayerToMap(new GeoJsonLayer(googleMap, new JSONObject(FileUtils.readFile(MapActivity.this, "vietnamsoilmap.geojson"))));
            addVietnamGeoJsonLayerToMap(new GeoJsonLayer(googleMap, new JSONObject(FileUtils.readFile(MapActivity.this, "vietnam.geojson"))));
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btnBack:
                onBackPressed();
                break;
            case R.id.tvDetail:
                Intent intent = new Intent(this, SoilDetailActivity.class);
                intent.putExtra("soil_id", soilSelected.getSoilId());
                startActivity(intent);
                break;
        }
    }

    private void moveCameraMap(double lat, double lon) {
        LatLng latLng = new LatLng(lat, lon);
        if (googleMap != null) {
            googleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(latLng, Constant.MapEnum.HANOI_ZOOM));
        }
    }

    @Override
    public void onMapClick(LatLng latLng) {

    }

    @Override
    public boolean onMarkerClick(Marker marker) {
        return false;
    }

    @Override
    public void onGetListSoilSuccess(ArrayList<SoilModel> soilModels) {
        this.soilModels = soilModels;
        expandLayout.expand();
        mAdapter = new MapAdapter(this, soilModels, this);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this, RecyclerView.VERTICAL, false);
        recycler_view_soil_map.setLayoutManager(layoutManager);
        recycler_view_soil_map.setItemAnimator(new DefaultItemAnimator());
        recycler_view_soil_map.setAdapter(mAdapter);
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public void onSoilMapClickListener(SoilModel soilModel) {
        soilSelected = soilModel;
        tvName.setText(soilModel.getNameVi());
        tvProvince.setText(provinceSelected);
        tvDes.setText(soilModel.getDescription());
        if (!TextUtils.isEmpty(soilModel.getUrl())) {
            Glide.with(MapActivity.this).load(soilModel.getUrl()).listener(new RequestListener<Drawable>() {
                @Override
                public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                    progressBar.setVisibility(View.GONE);
                    return false;
                }

                @Override
                public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                    progressBar.setVisibility(View.GONE);
                    return false;
                }

            }).into(imgSoil);

        }
        expandDetail.expand();
    }

    private void addColorsToMarkers(GeoJsonLayer layer) {
        // Iterate over all the features stored in the layer
        for (GeoJsonFeature feature : layer.getFeatures()) {

            GeoJsonPolygonStyle geoJsonPolygonStyle = new GeoJsonPolygonStyle();
            geoJsonPolygonStyle.setStrokeColor(Color.parseColor("#ff0000"));
            geoJsonPolygonStyle.setFillColor(getResources().getColor(Constant.getColorArea(feature.getProperty("domsoil").toLowerCase())));
            geoJsonPolygonStyle.setStrokeWidth(0f);
            feature.setPolygonStyle(geoJsonPolygonStyle);
        }
    }

    private void addGeoJsonLayerToMap(GeoJsonLayer layer) {
        addColorsToMarkers(layer);
        layer.addLayerToMap();
        // Demonstrate receiving features via GeoJsonLayer clicks.
        layer.setOnFeatureClickListener((GeoJsonLayer.GeoJsonOnFeatureClickListener) feature -> {

        });
    }

    //Layer tỉnh Vietnam
    private void addStrokeArea(GeoJsonLayer layer) {
        // Iterate over all the features stored in the layer
        for (GeoJsonFeature feature : layer.getFeatures()) {
            GeoJsonPolygonStyle geoJsonPolygonStyle = new GeoJsonPolygonStyle();
            geoJsonPolygonStyle.setStrokeColor(Color.parseColor("#ff0000"));
            geoJsonPolygonStyle.setStrokeWidth(2f);
            geoJsonPolygonStyle.setClickable(true);
            feature.setPolygonStyle(geoJsonPolygonStyle);
        }
    }

    private void addVietnamGeoJsonLayerToMap(GeoJsonLayer layer) {
        addStrokeArea(layer);
        layer.addLayerToMap();
        // Demonstrate receiving features via GeoJsonLayer clicks.
        layer.setOnFeatureClickListener((GeoJsonLayer.GeoJsonOnFeatureClickListener) feature -> {
            CommonUtils.showError(MapActivity.this, "Tỉnh " + feature.getProperty("name"));
            presenter.getListSoil("");
        });
    }
}

File: app/src/main/java/com/yenvth/soilDetectionApp/map/MapActivity.kt
package com.yenvth.soilDetectionApp.map

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.Spinner
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.LayoutManager
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.maps.*
import com.google.android.gms.maps.GoogleMap.OnMapClickListener
import com.google.android.gms.maps.GoogleMap.OnMarkerClickListener
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.maps.android.data.Feature
import com.google.maps.android.data.geojson.GeoJsonLayer
import com.google.maps.android.data.geojson.GeoJsonLayer.GeoJsonOnFeatureClickListener
import com.google.maps.android.data.geojson.GeoJsonPolygonStyle
import com.yenvth.soilDetectionApp.R
import com.yenvth.soilDetectionApp.databinding.ActivityMapBinding
import com.yenvth.soilDetectionApp.map.MapAdapter.OnSoilMapItemClickListener
import com.yenvth.soilDetectionApp.models.ProvinceModel
import com.yenvth.soilDetectionApp.models.SoilModel
import com.yenvth.soilDetectionApp.soilDetail.SoilDetailActivity
import com.yenvth.soilDetectionApp.utils.CommonUtils
import com.yenvth.soilDetectionApp.utils.Constant
import com.yenvth.soilDetectionApp.utils.Constant.MapEnum
import com.yenvth.soilDetectionApp.utils.FileUtils
import com.yenvth.soilDetectionApp.utils.ImageUtils
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class MapActivity : AppCompatActivity(), OnMapReadyCallback, View.OnClickListener,
    OnMapClickListener, OnMarkerClickListener, OnSoilMapItemClickListener {
    private lateinit var binding: ActivityMapBinding
    private val viewModel by viewModels<MapViewModel>()

    private val fusedLocationProviderClient: FusedLocationProviderClient? = null
    private var googleMap: GoogleMap? = null
    private var mUiSettings: UiSettings? = null
    private val mAdapter: MapAdapter by lazy {
        MapAdapter(this, this)
    }
    private var soilModels: ArrayList<SoilModel>? = null
    private var soilSelected: SoilModel? = null
    private var provinceModels: List<ProvinceModel> = arrayListOf()
    private var provinceSelected: ProvinceModel? = null
    private var provinceAdapter: ArrayAdapter<*>? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMapBinding.inflate(layoutInflater)
        setContentView(binding.root)
        supportActionBar?.hide()
        init()
        action()
        setupObserve()
    }

    @SuppressLint("NotifyDataSetChanged")
    private fun init() {
        viewModel.getProvinces()

        val mapFragment = supportFragmentManager.findFragmentById(R.id.map) as SupportMapFragment?
        mapFragment?.getMapAsync(this)

        val layoutManager: LayoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
        binding.recyclerViewSoilMap.layoutManager = layoutManager
        binding.recyclerViewSoilMap.itemAnimator = DefaultItemAnimator()
        binding.recyclerViewSoilMap.adapter = mAdapter
        mAdapter.notifyDataSetChanged()
    }

    private fun action() {
        binding.btnBack.setOnClickListener(this)
        binding.tvDetail.setOnClickListener(this)
        //Todo: Setup spinner
        provinceAdapter =
            ArrayAdapter(this, R.layout.layout_custom_spinner_not_icon, provinceModels)
        provinceAdapter?.setDropDownViewResource(R.layout.item_spinner)
        binding.spProvince.adapter = provinceAdapter
        binding.spProvince.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>?, view: View, i: Int, l: Long) {
                provinceSelected = provinceModels[i]
//                CommonUtils.showSnackBar(this@MapActivity, provinceSelected.provinceName(), false)
//                presenter.getListSoilByProvince(provinceSelected.provinceId())
            }

            override fun onNothingSelected(adapterView: AdapterView<*>?) {}
        }
    }

    private fun setupObserve() {
        viewModel.provinceSoil.observe(this) { soilModels ->
            mAdapter.setData(soilModels)
            binding.expandLayout.expand()
            binding.expandDetail.collapse()
        }

        viewModel.provinces.observe(this) { provinces ->
            this.provinceModels = provinces
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        this.googleMap = googleMap
        moveCameraMap(MapEnum.HANOT_LAT, MapEnum.HANOT_LON)
        googleMap.setOnMapClickListener(this)
        mUiSettings = googleMap.uiSettings
        mUiSettings?.isZoomControlsEnabled = false
        mUiSettings?.isMyLocationButtonEnabled = true
        mUiSettings?.isZoomGesturesEnabled = false
        mUiSettings?.isTiltGesturesEnabled = false
        mUiSettings?.isRotateGesturesEnabled = false
        mUiSettings?.isMapToolbarEnabled = false
        mUiSettings?.isScrollGesturesEnabled = false
        try {
            addGeoJsonLayerToMap(
                GeoJsonLayer(
                    googleMap,
                    JSONObject(FileUtils.readFile(this@MapActivity, "vietnamsoilmap.geojson"))
                )
            )
            addVietnamGeoJsonLayerToMap(
                GeoJsonLayer(
                    googleMap,
                    JSONObject(FileUtils.readFile(this@MapActivity, "vietnam.geojson"))
                )
            )
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.btnBack -> onBackPressed()
            R.id.tvDetail -> {
                val intent = Intent(this, SoilDetailActivity::class.java)
                intent.putExtra("soilId", soilSelected?.soilId)
                startActivity(intent)
            }
        }
    }

    private fun moveCameraMap(lat: Double, lon: Double) {
        val latLng = LatLng(lat, lon)
        if (googleMap != null) {
            googleMap?.animateCamera(
                CameraUpdateFactory.newLatLngZoom(
                    latLng,
                    MapEnum.HANOI_ZOOM.toFloat()
                )
            )
        }
    }

    override fun onMapClick(latLng: LatLng) {}
    override fun onMarkerClick(marker: Marker): Boolean {
        return false
    }

    override fun onSoilMapClickListener(soilModel: SoilModel) {
        soilSelected = soilModel
        binding.tvName.text = soilModel.nameVi
        binding.tvProvince.text = provinceSelected?.provinceName
        binding.tvDes.text = soilModel.description
        ImageUtils.setImage(this, soilModel.url, binding.imgSoil, binding.progressBar)
        binding.expandDetail.collapse()
        binding.expandDetail.expand()
    }

    private fun addColorsToMarkers(layer: GeoJsonLayer) {
        for (feature in layer.features) {
            val geoJsonPolygonStyle = GeoJsonPolygonStyle()
            geoJsonPolygonStyle.strokeColor = Color.parseColor("#ff0000")
            geoJsonPolygonStyle.fillColor = resources.getColor(
                Constant.getColorArea(
                    feature.getProperty("domsoil").toLowerCase(Locale.getDefault())
                )
            )
            geoJsonPolygonStyle.strokeWidth = 0f
            feature.polygonStyle = geoJsonPolygonStyle
        }
    }

    private fun addGeoJsonLayerToMap(layer: GeoJsonLayer) {
        addColorsToMarkers(layer)
        layer.addLayerToMap()
        layer.setOnFeatureClickListener(GeoJsonOnFeatureClickListener { feature: Feature? -> })
    }

    //Layer tỉnh Vietnam
    private fun addStrokeArea(layer: GeoJsonLayer) {
        for (feature in layer.features) {
            val geoJsonPolygonStyle = GeoJsonPolygonStyle()
            geoJsonPolygonStyle.strokeColor = Color.parseColor("#ff0000")
            geoJsonPolygonStyle.strokeWidth = 3f
            geoJsonPolygonStyle.isClickable = true
            feature.polygonStyle = geoJsonPolygonStyle
        }
    }

    private fun addVietnamGeoJsonLayerToMap(layer: GeoJsonLayer) {
        addStrokeArea(layer)
        layer.addLayerToMap()
        layer.setOnFeatureClickListener(GeoJsonOnFeatureClickListener { feature: Feature ->
            CommonUtils.showSnackBar(
                this@MapActivity,
                feature.getProperty("name") + " " + getString(R.string.province)
            )
            val provinceId = feature.getProperty("id_1").toIntOrNull()
            provinceSelected = provinceModels.find { it.provinceId == provinceId }
            binding.spProvince.setSelection(
                getIndex(
                    binding.spProvince,
                    provinceSelected?.provinceName ?: ""
                )
            )
            provinceId?.let { viewModel.getListSoilByProvince(it) }
        })
    }

    //private method of your class
    private fun getIndex(spinner: Spinner, myString: String): Int {
        for (i in 0 until spinner.count) {
            if (spinner.getItemAtPosition(i).toString().equals(myString, ignoreCase = true)) {
                return i
            }
        }
        return -1
    }
}